src/c++/vdo/base/data-vio.c

```diff
@@ -647,10 +650,26 @@ static void assign_discard_permit(struct limiter *limiter)
 	bio_list_add(limiter->permitted_waiters, bio);
 }
 
+#ifndef VDO_UPSTREAM
+#undef VDO_USE_ALTERNATE
+#if defined(RHEL_RELEASE_CODE) && defined(RHEL_MINOR) && (RHEL_MINOR < 50)
+#if (RHEL_RELEASE_CODE < RHEL_RELEASE_VERSION(10, 0))
+#define VDO_USE_ALTERNATE
+#endif
+#else /* !RHEL_RELEASE_CODE */
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(6, 10, 0))
+#define VDO_USE_ALTERNATE
+#endif
+#endif /* !RHEL_RELEASE_CODE */
+#endif /* !VDO_UPSTREAM */
 static void get_waiters(struct limiter *limiter)
 {
+#ifdef VDO_USE_ALTERNATE
 	bio_list_merge(&limiter->waiters, &limiter->new_waiters);
 	bio_list_init(&limiter->new_waiters);
+#else
+	bio_list_merge_init(&limiter->waiters, &limiter->new_waiters);
```

Comment:
I don't think we can do this; the point of the #ifdef is to be able to build on machines with an older kernel, and the older kernels simply don't have this function yet.

Reply:
Right, I meant something like 
 #ifdef VDO_ALTERNATIVE
 static inline bio_list_merge_init(...) { ... } 
 #endif

in our kernel code and then we can call it unconditionally. So on older kernels, we're just defining the function ourselves for our own private use.

Reply:
Is that really useful, though? In that case I think I'd have to conditionalize the userspace implementation to _not_ define the function so we don't get double-prototype problems. And we're still #ifdefing for different kernels so it doesn't simplify the kernel code much, either.

Reply:
Yeah, it's just a question of putting the messy ifdefs in the call site vs putting the messy ifdefs elsewhere and keeping the call site simple. Not a huge difference but I thought I'd at least raise the question.

As for the userspace version, defining the driver version if VDO_USE_ALTERNATIVE||!defined(__KERNEL__) and dropping the user-mode only version would probably work, or wrap the in-driver version in #ifndef __KERNEL__ and keep the user-mode only version too, since we'll need to keep a user-mode version for the long term anyway.

Reply:
Sorry that should've been (underscore)(underscore)KERNEL(underscore)(underscore) ... silly github

Reply:
Maybe I don't understand what you're talking about, but I don't see any version of this that doesn't lead to #ifdefs being more complicated and touching more files. There's a version of this that kind of makes sense if we were calling  bio_list_merge_init() from several files, but it's only called in two places (at least in our published code).